ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java

```diff
@@ -1164,13 +1195,49 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
         throw new InvalidColumnReferenceException("Field " + columnName + " is ambiguous.");
       }
     } else {
-      Expression baseExpression =
-          new QualifiedNameReference(
-              getLocation(context),
-              QualifiedName.of(dataSourceExtractor.getFromAlias())
-          );
-      return new DereferenceExpression(getLocation(context), baseExpression, columnName);
+      if (isAmbiguousColumnName(
+          columnName,
+          suffixFieldName,
+          dataSourceExtractor.getFromName(),
+          dataSourceExtractor.getFromAlias(),
+          dataSourceExtractor.getFromSchema())) {
+        throw new KsqlException(String.format("Ambiguous column name %s. You have stream/table"
+            + " with the same name/alias. Use stream/table name or"
+            + " alias as prefix when there is a column with the same name.", columnName));
+      }
+      if (columnName.equalsIgnoreCase(dataSourceExtractor.getFromAlias())
+          || columnName.equalsIgnoreCase(dataSourceExtractor.getFromName())) {
+        return new QualifiedNameReference(
+            getLocation(context),
+            QualifiedName.of(columnName)
+        );
+      } else {
+        Expression baseExpression =
+            new QualifiedNameReference(
+                getLocation(context),
+                QualifiedName.of(dataSourceExtractor.getFromAlias())
+            );
+        return new DereferenceExpression(getLocation(context), baseExpression, columnName);
+      }
+    }
+  }
+
+  private boolean isAmbiguousColumnName(
+      final String columnName,
+      final String suffixName,
+      final String sourceName,
+      final String sourceAlias,
+      final Schema sourceSchema
+  ) {
+    boolean isASchemaColumn = sourceSchema.fields().stream()
+        .anyMatch(field -> columnName.equalsIgnoreCase(field.name()));
+    if (suffixName == null) {
```

Comment:
can't we just omit this case? If there is no dereference there is certainly no ambiguity.

Reply:
Consider this example:
`CREATE STREAM S (S STRUCT<S INTEGER>) ...;`
`SELECT S FROM S;`
As you see we have ambiguity of the column name is the same as the stream name/alias.

Reply:
> SELECT S FROM S;

Is there any ambiguity here? Does not the `S` in the select always have to be the struct? It can't be the stream S or the struct's field S...

Reply:
No ambiguity but the idea is to be consistent in the way we treat the cases where we have column name the same as stream name/alias and require user to always provide stream name/alias in referring to such columns.